### PR TITLE
Scripts to update to .NET prerelease version

### PR DESCRIPTION
**Pull request by adityapatwardhan**
# PR Summary

Script to help to move to the next preview version of .NET. The version of .NET is in DotnetRuntimeMetadata.json.
The script is to be used by developer who wants to move the code base to the new .NET preview. The script will update the global.json and all the other csproj files which need to be updated.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: nuget.config [hunk 1]
```diff
@@ -2,6 +2,7 @@
 <configuration>
   <packageSources>
     <clear />
+    <add key="dotnet5" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/index.json" />
```
Comment (iSazonov) [inline nuget.config:5]: Does this allow anonymous access?

File: UpdateDotnetRuntime.ps1 [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+        $version = $_.Value.Version
+
+        foreach ($p in $pkgs) {
+            if ($p.Version -like "$versionPattern*") {
+                if ([System.Management.Automation.SemanticVersion] ($version) -lt [System.Management.Automation.SemanticVersion] ($p.Version)) {
+                    $_.Value.NewVersion = $p.Version
+                    break
+                }
+            }
```
Comment (iSazonov) [inline UpdateDotnetRuntime.ps1:67]: I think it is more simple to add new property in PowerShell.Common.props file and use the property in csproj files.

Comment (adityapatwardhan) [reply UpdateDotnetRuntime.ps1:67]: The versions for different packages can be different. We would need to add a property per package. Also the intention of this script is to be run manually before opening a PR to update .NET.

File: nuget.config (hunk 1 above)
Comment (adityapatwardhan) [reply nuget.config:5]: Yes, it is a public feed

File: UpdateDotnetRuntime.ps1 [hunk 3]
```diff
@@ -0,0 +1,130 @@
+[CmdletBinding()]
```
Comment (daxian-dbw) [inline UpdateDotnetRuntime.ps1:1]: Maybe this file should be put in `tools` folder?

**Review by daxian-dbw — COMMENTED** [review]
@adityapatwardhan Can you please update the PR description to include some information about what the script does and the context of why we need it?

File: UpdateDotnetRuntime.ps1 (hunk 3 above)
Comment (adityapatwardhan) [reply UpdateDotnetRuntime.ps1:1]: Makes sense.. I can move it.

**Review by iSazonov — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@daxian-dbw Ping..

File: tools/UpdateDotnetRuntime.ps1 [hunk 4]
```diff
@@ -0,0 +1,130 @@
+[CmdletBinding()]
```
Comment (TravisEz13) [inline tools/UpdateDotnetRuntime.ps1:1]: missing copyright header

File: tools/UpdateDotnetRuntime.ps1 [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+param (
+)
+
+function Update-GlobalJson([string] $Version) {
```
Comment (TravisEz13) [inline tools/UpdateDotnetRuntime.ps1:5]: a short description of each function would be good

File: tools/UpdateDotnetRuntime.ps1 [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+function Update-CsprojFile([string] $path, $values) {
+    $fileContent = Get-Content $path -raw
+    $updated = $false
+
+    foreach ($v in $values) {
+        if ($v.NewVersion) {
+            $stringToReplace = "<PackageReference Include=`"$($v.Name)`" Version=`"$($v.Version)`" />"
+            $newString = "<PackageReference Include=`"$($v.Name)`" Version=`"$($v.NewVersion)`" />"
+
+            $fileContent = $fileContent -replace $stringToReplace, $newString
```
Comment (TravisEz13) [inline tools/UpdateDotnetRuntime.ps1:87]: This is doing a regex replace... maybe it's safer to using <string>.replace()

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: tools/UpdateDotnetRuntime.ps1 (hunk 4 above)
Comment (adityapatwardhan) [reply tools/UpdateDotnetRuntime.ps1:1]: Fixed

File: tools/UpdateDotnetRuntime.ps1 (hunk 6 above)
Comment (adityapatwardhan) [reply tools/UpdateDotnetRuntime.ps1:87]: Regex replace is turning out to be a bit complicated. Keeping it as is. The intention is to create a PR after the script does the changes. So there will be a code review as a safety net.

File: tools/UpdateDotnetRuntime.ps1 (hunk 5 above)
Comment (adityapatwardhan) [reply tools/UpdateDotnetRuntime.ps1:11]: fixed

**Comment by adityapatwardhan** [comment]
@TravisEz13 please re-review

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**